be/dtm0_log.c

```diff
@@ -0,0 +1,631 @@
+			    struct m0_be_tx		 *tx,
+			    const struct m0_dtm0_tx_desc *txd,
+			    struct m0_buf		 *payload,
+			    struct m0_dtm0_log_rec	 *rec)
+{
+	bool		  is_persistent = log->dl_is_persistent;
+	struct m0_buf    *lpayload         = &rec->dlr_payload;
+	struct m0_be_seg *seg	           = log->dl_seg;
+
+	M0_PRE(m0_dtm0_log_rec__invariant(rec));
+
+	/* Attach payload to log if it is not attached */
+	if (!m0_buf_is_set(lpayload) && m0_buf_is_set(payload)) {
+		if (is_persistent) {
+			lpayload->b_nob = payload->b_nob;
+			M0_BE_ALLOC_BUF_SYNC(lpayload, seg, tx);
+			m0_buf_memcpy(lpayload, payload);
+			M0_BE_TX_CAPTURE_BUF(seg, tx, lpayload);
+			M0_BE_TX_CAPTURE_PTR(seg, tx, lpayload);
+		} else {
+			m0_buf_copy(lpayload, payload);
+		}
+	}
+
+	m0_dtm0_tx_desc_apply(&rec->dlr_txd, txd);
+	M0_POST(m0_dtm0_log_rec__invariant(rec));
+	if (is_persistent) {
+		M0_BE_TX_CAPTURE_ARR(seg, tx, rec->dlr_txd.dtd_ps.dtp_pa,
+					      rec->dlr_txd.dtd_ps.dtp_nr);
+	}
+
+	return 0;
+}
+
+M0_INTERNAL int m0_be_dtm0_log_update(struct m0_be_dtm0_log  *log,
+                                      struct m0_be_tx        *tx,
+                                      struct m0_dtm0_tx_desc *txd,
+                                      struct m0_buf          *payload)
+{
+	struct m0_dtm0_log_rec	*rec;
+
+	M0_PRE(payload != NULL);
+	M0_PRE(m0_be_dtm0_log__invariant(log));
+	M0_PRE(m0_dtm0_tx_desc__invariant(txd));
+	M0_PRE(m0_mutex_is_locked(&log->dl_lock));
+
+	return (rec = m0_be_dtm0_log_find(log, &txd->dtd_id)) ?
+                be_dtm0_log__set(log, tx, txd, payload, rec) :
+                m0_be_dtm0_log__insert(log, tx, txd, payload);
+}
+
+M0_INTERNAL int m0_be_dtm0_log_prune(struct m0_be_dtm0_log    *log,
+                                     struct m0_be_tx          *tx,
+                                     const struct m0_dtm0_tid *id)
+{
+	/* This assignment is meaningful as it covers the empty log case */
+	int                     rc = M0_DTS_LT;
+	struct m0_dtm0_log_rec *rec;
+	struct m0_dtm0_log_rec *currec;
+
+	M0_PRE(m0_be_dtm0_log__invariant(log));
+	M0_PRE(m0_dtm0_tid__invariant(id));
+	M0_PRE(m0_mutex_is_locked(&log->dl_lock));
+
+	m0_tl_for (lrec, log->dl_tlist, rec) {
+		if (!m0_dtm0_tx_desc_state_eq(&rec->dlr_txd,
+					      M0_DTPS_PERSISTENT))
+			return M0_ERR(-EPROTO);
+
+		rc = m0_dtm0_tid_cmp(log->dl_cs, &rec->dlr_txd.dtd_id, id);
+		if (rc != M0_DTS_LT)
+			break;
+	} m0_tl_endfor;
+
+	if (rc != M0_DTS_EQ)
+		return -ENOENT;
+
+	while ((currec = lrec_tlist_pop(log->dl_tlist)) != rec) {
+		M0_ASSERT(m0_dtm0_log_rec__invariant(currec));
+		be_dtm0_log_rec_fini(&currec, tx);
+	}
+
+	be_dtm0_log_rec_fini(&currec, tx);
+	return rc;
+}
+
+M0_INTERNAL void m0_be_dtm0_log_clear(struct m0_be_dtm0_log *log)
+{
+	struct m0_dtm0_log_rec *rec;
+
+	/* This function is expected to be called only on the client side where
+	   the log will always be a volatile log. */
+	M0_ASSERT(!log->dl_is_persistent);
```

Comment:
Maybe a better name for this "dl_is_persistent" is "dl_in_memory_only".
Otherwise, this "dl_is_persistent"might be confusing with the "M0_DTPS_PERSISTENT" state.

Reply:
or, we can rename it as "dl_is_volatile".

Reply:
Noted. I will add a TDB in code and visit this when we undertake a rewrite/refactoring of this code.